sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -952,63 +871,34 @@ public final FileResult close() throws Exception {
   }
 
   /**
-   * Result of a single bundle write. Contains the filename of the bundle.
+   * Result of a single bundle write. Contains the filename produced by the bundle, and if known
+   * the final output filename.
    */
-  public static final class FileResult {
-    private final String filename;
-    private final String destinationFilename;
+  public static final class FileResult implements Serializable {
+    private final ResourceId filename;
+    @Nullable private final ResourceId destinationFilename;
 
-    public FileResult(String filename, String destinationFilename) {
+    public FileResult(ResourceId filename, @Nullable ResourceId destinationFilename) {
       this.filename = filename;
       this.destinationFilename = destinationFilename;
     }
 
-    public String getFilename() {
+    public ResourceId getFilename() {
       return filename;
     }
 
-    public String getDestinationFilename() {
+    @Nullable public ResourceId getDestinationFilename() {
       return destinationFilename;
     }
 
-  }
-
-  /**
-   * A coder for FileResult objects.
-   */
-  public static final class FileResultCoder extends CustomCoder<FileResult> {
```

Comment:
Great question! 

"Update" is a feature of the Dataflow runner, that actually predates Beam. It allows updating a running streaming pipeline with new code, even allowing certain changes to the graph structure. This has not yet been proposed as part of the Beam model, though it could be and should be.

While update is not yet formally a part of Beam, we should endeavor not to gratuitously break runner features - especially features that are widely used. Similarly Flink supports resuming a pipeline from a savepoint, and we should try not to break that.

In this particular case "update compatibility" (which means ensuring that a new graph structure can resume from an old graph) is not the issue. The issue is that the Java Serializable structure for objects is extremely prone to changing. The user might make simple code changes that do not at all change the graph structure of the pipeline, yet will change the serialized format of the object. This means that the new pipeline will start crashing at runtime, because it will no longer be able to deserialize it's own stored data.

Supporting undocumented runner features is not sustainable. We really need to make sure all such things are explicitly captured in the Beam model.